backend/svirt.pm

```diff
@@ -167,25 +171,33 @@ sub save_snapshot {
 
 sub load_snapshot {
     my ($self, $args) = @_;
-    my $snapname = $args->{name};
-    my $vmname   = $self->console('svirt')->name;
-    my $rsp;
+    my $snapname                   = $args->{name};
+    my $vmname                     = $self->console('svirt')->name;
     my $post_load_snapshot_command = '';
+    my $vm_state;
+    my $rsp;
     if (check_var('VIRSH_VMM_FAMILY', 'hyperv')) {
         my $ps = 'powershell -Command';
+        $self->run_ssh_cmd("$ps (Get-VMCheckpoint -VMName $vmname).Name");
         $rsp = $self->run_ssh_cmd(qq($ps "\$ProgressPreference='SilentlyContinue'; Restore-VMSnapshot -VMName $vmname -Name $snapname -Confirm:\$false"));
+        # after restoring from checkpoint, VM should be 'Running'.
+        (undef, $vm_state) = $self->run_ssh_cmd("$ps (Get-VM -VMName $vmname).State", wantarray => 1);
+        $vm_state = trim($vm_state);
+        bmwqemu::diag "VM state: $vm_state, after restoring from checkpoint";
         $self->run_ssh_cmd("mv -v xfreerdp_${vmname}_stop xfreerdp_${vmname}_stop.bkp", $self->get_ssh_credentials('hyperv'));
-
-        for my $i (1 .. 5) {
-            # Because of FreeRDP issue https://github.com/FreeRDP/FreeRDP/issues/3876,
-            # we can't connect too "early". Let's have a nap for a while.
-            sleep 10;
-            last
-              unless $self->run_ssh_cmd(
-                "pgrep --full --list-full xfreerdp.*\$(cat xfreerdp_${vmname}_stop.bkp)",
-                $self->get_ssh_credentials('hyperv'));
-            $self->die("xfreerdp did not start") if ($i eq 5);
-        }
+        # Because of FreeRDP issue https://github.com/FreeRDP/FreeRDP/issues/3876,
+        # we can't connect too "early". Let's have a nap for a while.
+        my $i = 0;
+        do {
+            sleep 5;
+            (undef, $vm_state) = $self->run_ssh_cmd("$ps (Get-VM -VMName $vmname).State", wantarray => 1);
+            $vm_state = trim($vm_state);
+            bmwqemu::diag "Restoring RDP connection, attempts: ${\++$i}, VM state: $vm_state";
+        } while ($i < 5 && $self->run_ssh_cmd("pgrep --full --list-full xfreerdp.*\$(cat xfreerdp_${vmname}_stop.bkp)", $self->get_ssh_credentials('hyperv'))
+            && $vm_state eq 'Running');
+
+        $vm_state eq 'Running' or $self->die("The VM is not running");
+        $self->die("xfreerdp did not start") if ($i == 5);
```

Comment:
I'd avoid adding magic numbers like 5, especially it it occurs multiple times it would make sense to use a constant. You might also want to default-initialize the constant from a test or environment variable to make this configurable.

Reply:
Indeed, that makes perfect sense, thanks!